feat!: upgraded python to 3.12 and django to 4.2

huniafatima-arbi opened the pull request:
**Merge checklist:**
Check off if complete *or* not applicable:
- [x] Documentation updated (not only docstrings)
- [x] Fixup commits are squashed away
- [x] Unit tests added/updated
- [ ] Manual testing instructions provided
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

**Post-merge:**
- [ ] [Backported](https://openedx.atlassian.net/wiki/spaces/COMM/pages/2065367719/Making+a+pull+request+for+a+named+release) to latest and next named releases

UsamaSadiq reviewed (commented):
Changes look good but let's test the docker build locally with the updated Dockerfile to make sure it doesn't break the pipeline after deployment.

aht007 commented on .github/workflows/docker-publish.yml:
```diff
@@ -4,12 +4,14 @@ on:
   push:
     branches:
       - main
+  pull_request:
```
Is this change temporary?

huniafatima-arbi replied:
yes, it's there to test the docker image build on CI

aht007 commented on .github/workflows/docker-publish.yml:
```diff
@@ -19,4 +21,4 @@ jobs:
           DOCKERHUB_PASSWORD: ${{ secrets.DOCKERHUB_PASSWORD }}
           DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
-        run : make github_docker_push
+        run : make github_docker_build
```
Also why is this change needed?

huniafatima-arbi replied:
This change is temporary as well as we didn't want to publish the testing image to docker hub

aht007 reviewed (commented):
Looks reasonable, I have added a couple of questions though. I have tested the service locally and things look good. I will give the testing phase another pass once the changes are on staging.

huniafatima-arbi commented:
> Looks reasonable, I have added a couple of questions though. I have tested the service locally and things look good. I will give the testing phase another pass once the changes are on staging.

Hi @aht007 , I have reverted the testing code. Please let me know if the PR is good to go.

aht007 commented on .github/workflows/docker-publish.yml:
```diff
@@ -4,12 +4,13 @@ on:
   push:
     runs-on: ubuntu-latest
-    if: github.event_name == 'push'
+    # if: github.event_name == 'push'
```
Is this still needed?

huniafatima-arbi replied:
resolved

aht007 reviewed (approved):
LGTM, lets test this on staging before rolling out to production.
